### PR TITLE
HPCC-30996 Fix issues with roxie copying from remote storage source

### PR DESCRIPTION
**Pull request by jakesmith**
Roxie deployment (via esp) creates new logical file entries in the local Dali and annotates it with cloneFrom* info, so that at runtime, Roxie can read remotely, and copy file parts asynchronously in the background to the local placements. This did not work where the remote file was from a non-standard topology, in particular where the remote files were from cloud, where parts are by default stored per directly, and where storage may be striped.
Fix by cloning the source file descriptor as a reference point, such that the remote locations and fully use the remote file descriptor as a source.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/dfu/dfuutil.cpp [hunk 1]
```diff
@@ -832,7 +845,14 @@ class CFileCloner
         }
         else
         {
+            IPropertyTree *dstClonedFDesc = dfile->queryAttributes().queryPropTree("cloneFromFDesc");
+            IPropertyTree *srcClonedFDesc = srcfdesc->queryProperties().queryPropTree("cloneFromFDesc");
+            if ((dstClonedFDesc && !srcClonedFDesc) || (!srcClonedFDesc && dstClonedFDesc) ||
+                (areMatchingPTrees(dstClonedFDesc, srcClonedFDesc)))
+                return true;
             StringBuffer s;
+            if (checkValueChanged(dfile->queryAttributes().queryProp("@cloneFrom"), srcdali->endpoint().getUrlStr(s).str()))
```
Comment (afishbeckln) [inline dali/dfu/dfuutil.cpp:854]: Is this check now duplicated?

**Review by afishbeckln — COMMENTED** [review]
Conceptually looks great.  One boolean check question and a duplicate line.

File: dali/dfu/dfuutil.cpp [hunk 2]
```diff
@@ -832,7 +845,14 @@ class CFileCloner
         }
         else
         {
+            IPropertyTree *dstClonedFDesc = dfile->queryAttributes().queryPropTree("cloneFromFDesc");
+            IPropertyTree *srcClonedFDesc = srcfdesc->queryProperties().queryPropTree("cloneFromFDesc");
+            if ((dstClonedFDesc && !srcClonedFDesc) || (!srcClonedFDesc && dstClonedFDesc) ||
+                (areMatchingPTrees(dstClonedFDesc, srcClonedFDesc)))
```
[the remark this answers is not in the document]
Comment (jakesmith) [reply dali/dfu/dfuutil.cpp:851]: whoops yes.

File: dali/dfu/dfuutil.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/dfu/dfuutil.cpp:854]: I guess yes, if both cloneFromFDesc's exists, should check them, but avoid other tests.
I'll change code.

Comment (jakesmith) [reply dali/dfu/dfuutil.cpp:854]: oh, line was duplicated!
Well it made me look at an optimization anyway

**Comment by jakesmith** [comment]
@afishbeckln - please review changes.

**Review by afishbeckln — APPROVED** [review]
Looks good.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**